#################################################
##       Script to check if primers overlap    
##          variants in patient genome  		
##	 Targeted resequencing on the MiSeq 				
##  Aparicio Lab WSOP 2013-001 developed by			
##	 Dr Damian Yap , Research Associate				
##	 dyap@bccrc.ca  Version 2.0 (Aug 2013) 			
##################################################

###############################################################
# Run as source("/home/dyap/R/R-scripts/Chk_SNP_primers.R")
# IF for the first time, use firstrow=1
firstrow=1
################################################################

###############
# REQUIREMENTS
##############

# Manifest file (from MiSeq)
# Manifest file (generated by UCSC in silico PCR)
# Variants (patient specific, sorted by chromosome) at the following location on beast
# /share/lustre/projects/breast_xeno_evolution/apolloh2.0/germlineVariant_SAMtools/SA429N/SA429N_germlineVarPosns_chr01.txt 
# These variants are in .vcf format as a txt file (filtered by Gavin)


##################
# These commands must be specifed in order for this script to work
# source("http://www.bioconductor.org/biocLite.R"); source("http://www.bioconductor.org/biocLite.R"); biocLite("BSgenome"); 
# biocLite("BSgenome.Hsapiens.UCSC.hg19"); library('BSgenome.Hsapiens.UCSC.hg19')
# install.packages("XLConnect")
# install.packages("XML", repos = "http://www.omegahat.org/R")
# CRAN (http://cran.r-project.org/)
# install.packages("audio")


library(Biostrings)
library("IRanges")
library("GenomicRanges")
library(Rsamtools)
library("VariantAnnotation")
library(foreign)

#################################################
# Directory structure and file names

basedir="/share/lustre/backup/dyap/Projects/MiSeq_Data"

sumdir="/home/dyap/Projects/Tumour_Evol/Summary"

manfile <- paste(sumdir, paste(sample,"Primers.csv", sep="_"), sep="/")
outfile <- paste(sumdir, paste(sample,"SNP_UCSCman_chk.csv", sep="_"), sep="/")

manifestdir=paste(basedir,"AmpliconManifests", sep="/")

######################################################


