#!/bin/sh

# This Script was written by Damian Yap (Apr 2013)
# To filter and add specific adaptors to primer3 generated primers for MiSeq
# Generate the order file and filtered AmpliconManifest

# Preparation of files
# cd /home/dyap/Projects/PrimerDesign/ITH3
#cp /share/lustre/fdorri/projects/primer_design/ith_patient_3_selected_1500_more_pos/2015-09-24_16-44-14/DG1135a_fdorri_ITH_selected_1500_more/outputs/intermediate/Final_Single_TASK_6_primer_list.txt DG1135a_primer_list.txt
#cp /share/lustre/fdorri/projects/primer_design/ith_patient_3_selected_1500_more_pos/2015-09-24_16-44-14/DG1135c_fdorri_ITH_selected_1500_more/outputs/intermediate/Final_Single_TASK_6_primer_list.txt DG1135c_primer_list.txt
#cp /share/lustre/fdorri/projects/primer_design/ith_patient_3_selected_1500_more_pos/2015-09-24_16-44-14/DG1135e_fdorri_ITH_selected_1500_more/outputs/intermediate/Final_Single_TASK_6_primer_list.txt DG1135e_primer_list.txt
#cp /share/lustre/fdorri/projects/primer_design/ith_patient_3_selected_1500_more_pos/2015-09-24_16-44-14/DG1135g_fdorri_ITH_selected_1500_more/outputs/intermediate/Final_Single_TASK_6_primer_list.txt DG1135g_primer_list.txt
###############################################################
for fname in DG1135a DG1135c DG1135e DG1135g
	do
#fname="DG1135a"	# Sample set naming
name="study_ith."$fname
expt="1"
type="Ill" 		# Illumina adaptors
#type="Fld" 		# Fluidigm adaptors
Project="manual"	# Project name
RefGen="Custom\\"$name	# Reference Genome for AmpliconManifest
ver="2"			# AmpliconManifest Version
###############################################################

dir="/home/dyap/Projects/PrimerDesign/ITH3"
p3dir=$dir
posdir=$dir

##########################################
# Selected positions should be placed here
# Generated de novo from file
sourcefile=$posdir"/selected_positions.txt"
ffile=$fname"_primer_list.txt"
# List the unique seq_id
cat $dir"/"$ffile | awk -F"," '{print $1}' | uniq > $sourcefile

#########################################
# Primer name format
# study_ith.patient_7.seqval_1.sample_DG1137A4
# patient_15.seqval_1
setname=$name".seqval_"$expt
#########################################
# Selects only the breakpoints (unique starts)
# grep breakpoint $dir"/"$ffile | awk -F"\t" '{print $17}' > $sourcefile

#clear 
cd $dir


primerfile=$dir"/"$ffile
orderleft=$p3dir"/"$setname"-"$type"-leftprimers.csv"
orderright=$p3dir"/"$setname"-"$type"-rightprimers.csv"

						echo "WellPosition,Name,Sequence,Notes" > $orderleft
						echo "WellPosition,Name,Sequence,Notes" > $orderright

#oldmanifest=$p3dir"/"$name".AmpliconManifest"
#newmanifest=$p3dir"/"$name"-filtered.AmpliconManifest"

rm -fr $orderfile
#rm -fr $newmanifest


	if [[ $type == "Ill" ]]
		then
		# Illlumina Adaptors (5'->3')
		fa="TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG"
		ra="GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAG"
	fi

	if [[ $type == "Fld" ]]
		then	
		# Forward adaptor for Fluidigm
		fa="ACACTGACGACATGGTTCTACA"
		# Reverse adaptor for Fluidigm (5'->3')
		ra="TACGGTAGCAGAGACTTGGTCT"
	fi

# If file has a header (manual input)
# cat $sourcefile | tail -n +2  > pattern
# If file has a NO header (selected using grep)
# cat $sourcefile > pattern
# IF file is generated by a script (header and formatted)
# cat $sourcefile | tail -n +2 | awk -F" " '{print $1}' > pattern
# IF file is generated by a script (NO header but formatted)
cat $sourcefile | awk -F"," '{print $1}' > pattern

# Counters
alpha="A"
count=1
plate=1
		for i in `cat pattern`

		do
		echo $i

# Check the positions of these in the input file ALWAYS!!!!

		label=`grep -m1 $i $primerfile | awk -F"," '{print $1}'`
		left0=`grep -m1 $i $primerfile | awk -F"," '{print $4}'`
		right0=`grep -m1 $i $primerfile | awk -F"," '{print $5}'`


		echo $label,$left0,$right0

				if [[ $alpha == "I" ]]; 
					then 	alpha="A"
						plate=$((plate+1))
						orderleft=$p3dir"/"$setname"-L"$plate".csv"
						orderright=$p3dir"/"$setname"-R"$plate".csv"

						echo "WellPosition,Name,Sequence,Notes" > $orderleft
						echo "WellPosition,Name,Sequence,Notes" > $orderright

				fi

			if [[ $left0 != "" ]];
			
				then

				if [[ $right0 != "" ]];
			
					then

						if [[ $count -lt 10 ]];
							then 
								welpos=$alpha"0"$count
							else
								welpos=$alpha$count
						fi

						echo "Adding adaptors to sequences..."
						echo $welpos","$label"_L,"$fa$left0 >> $orderleft
						echo $welpos","$label"_R,"$ra$right0 >> $orderright

#						echo "Filtering AmpliconManifest"
#						grep ^$i -m1 $oldmanifest | awk -F"\t" '{print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6}' | sed 's/chr//' >> $newmanifest.tmp

				fi
			fi

			count=$((count+1))

			if [[ $count == 13 ]];
				then 
				count="1"


				if [[ $alpha == "H" ]]; then alpha="I"; fi
				if [[ $alpha == "G" ]]; then alpha="H"; fi
				if [[ $alpha == "F" ]]; then alpha="G"; fi
				if [[ $alpha == "E" ]]; then alpha="F"; fi
				if [[ $alpha == "D" ]]; then alpha="E"; fi
				if [[ $alpha == "C" ]]; then alpha="D"; fi
				if [[ $alpha == "B" ]]; then alpha="C"; fi
				if [[ $alpha == "A" ]]; then alpha="B"; fi
				if [[ $alpha == "next" ]]; then alpha="A"; fi


			fi

		
		done

#                echo "[Header]" > $newmanifest
#                echo $name"  Manifest Version,"$ver >> $newmanifest
#                echo "ReferenceGenome,"$RefGen >> $newmanifest
#                echo "  " >> $newmanifest
#                echo "[Regions]" >> $newmanifest
#                echo "Name,Chromosome,Amplicon Start,Amplicon End,Upstream Probe Length,Downstream Probe Length,Comments" >> $newmanifest
#		cat $newmanifest.tmp >> $newmanifest



#		echo "Checking Amplicon Manifest for duplicates in the name column"

#		dup=`awk -F'\t' '{print $1}' $newmanifest.tmp | sort | uniq -d | grep -F -f - $newmanifest.tmp`

#			if [[ $dup != "" ]];
#				then
#				echo "DUPLICATES in AMPLICONMANIFEST"
#				echo $dup
#			fi
rm -f $p3dir/*.tmp
done			
echo "done."
exit;
